nova/compute/api.py

```diff
@@ -695,6 +695,33 @@ def _merge_bdms_lists(self, overridable_mappings, overrider_mappings):
                 [bdm for bdm in overridable_mappings
                  if bdm['device_name'] not in device_names])
 
+    def _flavor_enforces_boot_from_volume(self, flavor):
+        """Determines whether boot-from-volume is enforced by flavor"""
+        return (flavor.get('extra_specs', {})
+                .get('boot_from_volume', 'false') == 'true')
+
+    def _enforce_boot_from_volume(self, image_meta, flavor,
+                                  block_device_mapping):
+        """Adds boot-from-volume block device mapping if not already set."""
+        root_bdm = block_device.get_root_bdm(block_device_mapping)
+        if root_bdm:
+            if root_bdm.get('destination_type') == 'volume':
+                return
+            elif (root_bdm.get('source_type') == 'image' and
+                  root_bdm.get('destination_type') == 'local'):
```

Comment:
As Jakob pointed out, there could be a blank disk/swap disk added via `blank -> local`. We don't support ephemeral storage and thus also cannot support that case. Anything requesting `destination_type == 'local'` needs to be prohibited.